README.md
# HTTP Sink

A trivial HTTP Sinkhole. Always returns a 200 OK and JSON response body with
request details.

## Usage

### CLI

Start a sinkhole on port 8080 using npx:

```bash
# Listen on the default localhost:8080
npx httpsink
```

Alternatively install the module globally, then start a sinkhole:

```bash
npm install -g httpsink

# Listen on port 3000 on a public interface
httpsink --port 3000 --host '0.0.0.0'
```

### Docker via Pre-built Image

Exposes the sinkhole on port 8080:

```bash
docker run --rm -p 8080:8080 quay.io/evanshortiss/httpsink
```

You can specify a version tag if desired.

### Docker via npx

Uses npx to download the latest version of `httpsink`, and exposes the sinkhole on port 8080:

```bash
docker run -p 8080:8080 --rm --entrypoint /bin/sh node:14-alpine -c 'npx httpsink --host "0.0.0.0" --port 8080'
```


### Module

```js
const sinkhole = require('httpsink')

async function main () {
  // Create a sinkhole that listens on port 8080
  const server = await sinkhole({
    port: 8080
  })

  // ...sometime later call server.close() when you're done
  server.close()
}

main()
```

### Response Format

Each request receives the same response format:

```js
{
  // IP the server determines to be the origin of your request
  "ip": "127.0.0.1",

  // Unique ID generated for this request
  "uuid": "pDo-u-eOpxiVH_15aSV05",

  // When the server started processing the request
  "recv": "2021-02-04T19:25:03.740Z",

  // When the server sent the response
  "resp": "2021-02-04T19:25:03.744Z",

  // Bytes read for a request that sends a body
  "bodyBytesSize": 0

  // Bytes read for the entire HTTP transmission
  "totalBytesSize": 124

  // Method and URL that was requested
  "method":"POST",
  "url":"/json-test",

  // Hostname of the host (e.g a Pod hostname in Kubernetes)
  "hostname": "httpsink-85d65bfdd8-429kb",

  // Version of the httpsink module being run
  "version": "1.4.0"
}
```


CHANGELOG.md
# 1.4.0 (16/11/2022)

* Add `version` to response JSON payload.

# 1.3.0 (22/07/2022)

* Add `hostname` to response JSON payload.

# 1.2.0 (21/03/2022)

* Set `connection: close` header to speed up HTTP `server.close()`.

# 1.1.1 (05/02/2021)

* Fix shutdown log format.

# 1.1.1 (05/02/2021)

* Handle process kill signals.

# 1.1.0 (05/02/2021)

* Add to `totalBytesSize` and `ip` to response JSON.
* Set `content-type: application/json` in response headers.
* Rename `bytesRead` to `bodyBytesSize`.
* Rename `path` to `url`.

# 1.0.0 (04/02/2021)

* Initial release.


test/index.js
'use strict';

const test = require('tape');
const http = require('http');
const sinkhole = require('../index');
const { randomBytes } = require('crypto');

test('create a sinkhole listening on 127.0.0.1:8080', (t) => {
  sinkhole()
    .then((server) => {
      t.ok(server);
      t.equal(server.address().port, 8080);
      t.equal(server.address().address, '127.0.0.1');

      server.close((err) => t.end(err));
    })
    .catch((e) => t.end(e));
});

test('create a sinkhole listening on 0.0.0.0:3000', (t) => {
  const opts = {
    host: '0.0.0.0',
    port: 3000
  };

  sinkhole(opts)
    .then((server) => {
      t.ok(server);
      t.equal(server.address().port, opts.port);
      t.equal(server.address().address, opts.host);

      server.close((err) => t.end(err));
    })
    .catch((e) => t.end(e));
});

test('sinkhole response format and time calculation', (t) => {
  sinkhole()
    .then((server) => {
      const requestByteCount = 50;
      const requestDurationMs = 25;
      const { address, port } = server.address();
      console.log(`http://${address}:${port}/`);

      const request = http.request(
        {
          host: address,
          port,
          method: 'POST'
        },
        (res) => {
          let chunks = '';

          res.setEncoding('utf8');
          res.on('data', (chunk) => (chunks += chunk));
          res.on('end', () => {
            const {
              uuid,
              ip,
              recv,
              resp,
              bodyBytesSize,
              totalBytesSize,
              method,
              url,
              hostname,
              version
            } = JSON.parse(chunks);

            t.isEqual(res.headers['content-type'], 'application/json');
            t.isEqual(ip, '127.0.0.1');
            t.isEqual(method, 'POST');
            t.isEqual(url, '/');
            t.isEqual(typeof uuid, 'string');
            t.isEqual(typeof recv, 'string');
            t.isEqual(typeof resp, 'string');
            t.isEqual(typeof bodyBytesSize, 'number');
            t.isEqual(typeof totalBytesSize, 'number');
            t.isEqual(typeof hostname, 'string');
            t.isEqual(typeof version, 'string');

            // Verify the request duration was close to the specified duration
            // Timers are not accurate, so give 15% leeway. Yes, 15% variance!
            const time = new Date(resp).getTime() - new Date(recv).getTime();
            t.ok(time >= requestDurationMs * 0.85);

            // Bytes read should be equal to those sent
            t.isEqual(bodyBytesSize, requestByteCount);

            server.close((err) => t.end(err));
          });
        }
      );

      // Write a specific number of random bytes
      request.write(randomBytes(requestByteCount));

      // End the request and let the server respond after a delay
      setTimeout(() => request.end(), requestDurationMs);
    })
    .catch((e) => t.end(e));
});


index.js
'use strict';

const http = require('http');
const log = require('barelog');
const { nanoid } = require('nanoid');
const hostname = require('os').hostname();
const { version } = require('./package.json')

/**
 * @typedef {Object} SinkHoleOptions
 * @property {Number} port
 * @property {String} host
 */

/**
 * Initialise a sinkhole server with the given options.
 * @param {SinkHoleOptions} options
 * @returns {Promise<http.Server>}
 */
module.exports = (options) => {
  if (!options) {
    options = {
      host: 'localhost',
      port: 8080
    };
  }

  let { port = 8080, host = 'localhost' } = options;

  const server = http.createServer((req, res) => {
    const recv = new Date().toJSON();
    const uuid = nanoid();
    const { method, url } = req;
    let bodyBytesSize = 0;

    log(`(recv ${uuid}) ${req.method} ${req.url}`);

    req.on('data', (chunk) => {
      bodyBytesSize += Buffer.byteLength(chunk);
    });

    req.on('error', (e) => {
      log(`Error occurred for request - (${uuid}) ${req.method} ${req.url}:`);
      log(e);
    });

    req.on('end', () => {
      log(`(resp ${uuid}) ${req.method} ${req.url}`);
      res.setHeader('content-type', 'application/json');
      res.setHeader('connection', 'close');
      res.writeHead(200);
      res.end(
        JSON.stringify({
          ip: req.headers['x-forwarded-for'] || req.socket.remoteAddress,
          method,
          url,
          uuid,
          recv,
          resp: new Date().toJSON(),
          bodyBytesSize,
          totalBytesSize: req.socket.bytesRead,
          hostname,
          version
        })
      );
    });
  });

  return new Promise((resolve, reject) => {
    try {
      server.listen(port, host, () => resolve(server));
    } catch (e) {
      reject(e);
    }
  });
};
